src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -2,18 +2,52 @@
 # Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
 #
 
-from typing import List, Optional
+from typing import TYPE_CHECKING, DefaultDict, Dict, List, Optional, Tuple
 
-from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark._internal.analyzer.expression import Attribute, Expression, Star
 from snowflake.snowpark._internal.analyzer.snowflake_plan_node import Limit, LogicalPlan
+from snowflake.snowpark._internal.analyzer.unary_expression import UnresolvedAlias
+
+if TYPE_CHECKING:
+    from snowflake.snowpark._internal.analyzer.analyzer import Analyzer
+
+
+def infer_quoted_identifiers_from_expressions(
+    expressions: List[Expression],
+    analyzer: "Analyzer",
+    df_aliased_col_name_to_real_col_name: DefaultDict[str, Dict[str, str]],
+) -> Optional[List[str]]:
+    """
+    Infer quoted identifiers from (named) expressions.
+    """
+    from snowflake.snowpark._internal.analyzer.select_statement import parse_column_name
+    from snowflake.snowpark._internal.utils import quote_name
+
+    result = []
+    for e in expressions:
+        # If we do select *, we may not be able to get all current quoted identifiers
+        # (e.g., when SQL simplifier is disabled), so we just be conservative and do
+        # not perform any inference in this case.
+        if isinstance(e, UnresolvedAlias) and isinstance(e.child, Star):
+            return None
+        column_name = parse_column_name(
+            e, analyzer, df_aliased_col_name_to_real_col_name
+        )
+        if column_name is not None:
+            result.append(quote_name(column_name))
+        else:
+            return None
+    return result
 
 
 def infer_metadata(
     source_plan: LogicalPlan,
-) -> Optional[List[Attribute]]:
+    analyzer: "Analyzer",
+    df_aliased_col_name_to_real_col_name: DefaultDict[str, Dict[str, str]],
+) -> Tuple[Optional[List[Attribute]], Optional[List[str]]]:
```

Comment:
Instead of a tuple, let's get a NamedTuple Maybe call it PlanMetadata with attributes, and quoted_identifier

Reply:
done